File: src/test/java/hex/paymentstore/domain/AddPaymentUseCaseTest.java
package hex.paymentstore.domain;

import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import org.mockito.stubbing.Answer;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class AddPaymentUseCaseTest {
	private final PaymentRepository repository = Mockito.mock(PaymentRepository.class);
	private final ArgumentCaptor<Payment> argumentCaptor = ArgumentCaptor.forClass(Payment.class);
	private final AddPaymentUseCase sut = new AddPaymentUseCase(repository);
	
	@Test
	void should_add_payment() {
		//given
		when(repository.add(argumentCaptor.capture())).thenAnswer((Answer<Payment>) invocation -> argumentCaptor.getValue());
		Payment payment = new Payment("id", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123"));
		
		//when
		Payment added = sut.add(payment);
		
		//then
		assertEquals(added.id().length(), 20);
		assertEquals(added.amount(), payment.amount());
		assertEquals(added.currency(), payment.currency());
		assertEquals(added.userId(), payment.userId());
		assertEquals(added.account(), payment.account());
		
		verify(repository).add(added);
	}
}

File: src/test/kotlin/hex/paymentstore/domain/AddPaymentUseCaseTest.kt
package hex.paymentstore.domain

import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import java.math.BigDecimal

class AddPaymentUseCaseTest : ShouldSpec({
  val repository = mockk<PaymentRepository>()
  val slot = slot<Payment>()
  every { repository.add(capture(slot)) } answers {slot.captured}
  val sut = AddPaymentUseCase(repository)
 
 should("add payment") {
  //given
  val payment = Payment("id", BigDecimal.TEN, "PLN", UserId("userId"), BankAccountNumber("123"))
  
  //when
  val added = sut.add(payment)
 
  //then
  added.id!!.length shouldBe 20
  added.amount shouldBe payment.amount
  added.currency shouldBe payment.currency
  added.userId shouldBe payment.userId
  added.account shouldBe payment.account
  
  verify { repository.add(added) }
 }
})
